### ISSUE TITLE
wil::com_ptr typedef is silently omitted if WI_ENABLE_EXCEPTIONS isn't defined

### ISSUE THREAD

**Issue by giuntnic** [microsoft/wil#109]
If exceptions aren't enabled using wil::com_ptr results in an error that isn't immediately clear without digging through the source. 
`error C2039: 'com_ptr': is not a member of 'wil' `
It should be pretty easy to add a struct that takes a compile time message and static_assert(false, message) if that class is instantiated to produce a clearer error message when using throwing types in binaries that don't support exceptions.

**Comment by sylveon** [comment]
There's `wil::com_ptr_nothrow`

**Comment by giuntnic** [comment]
Yeah, I switched over to the _nothrow variant once I figured out what the error was. I made this change in a header that is included into a bunch of different binaries one of which happened to have exceptions turned off and it took me a bit to figure out that that binary didn't have exceptions and that was causing the compiler error since the typedef isn't defined at all in that case. Adding some sort of InvalidInstantiation struct to static_assert in cases like this would be pretty helpful in making the error immediately apparent. This is a little tricky since string literals can't be passed as template arguments but maybe something like this:
`
enum class BadInstantiation : int { com_ptr }
template<BadInstantiation>
struct InvalidInstantiation {
    static_assert(false, "Must be overloaded")
};
template<>
struct InvalidInstantiation<BadInstantiation::com_ptr> {
    static_assert(false, "wil::com_ptr must not be instantiated when exceptions are disabled.")
}
`
although admittedly that is still pretty ugly.

**Comment by dunhor** [comment]
What error was it giving you that was confusing? I'd expect it to give some form of "undefined symbol" error, which should be relatively clear as to what's wrong. Albeit perhaps not super clear of _why_ the symbol is undefined, however looking up the definition should give a pretty quick answer:
```c++
#ifdef WIL_ENABLE_EXCEPTIONS
    //! COM pointer, errors throw exceptions (see @ref com_ptr_t for details)
    template <typename T>
    using com_ptr = com_ptr_t<T, err_exception_policy>;
#endif
```
But if we _were_ to try and do something here, we can probably piggy-back off the definition of `err_exception_policy` which we already have to stub out for other reasons. E.g. providing functions of the form:
```c++
struct err_exception_policy
{
    template <typename T = void> static result HResult(HRESULT hr) { static_assert(details::false_v<T>, "Exceptions not enabled etc. etc."); }
    // ...
};
```
It's probably best to do that _anyway_ since the experience of `err_exception_policy` accidentally being used is currently pretty sub-optimal (although it should be somewhat difficult to find yourself in that situation).

But again, I'm not all that sold on this being necessary. Additionally a lot of people are generally unhappy if a symbol shows up in intellisense/editor suggestions that they can't use

**Comment by giuntnic** [comment]
Yeah, the what is fairly clear in `error C2039: 'com_ptr': is not a member of 'wil'` but tracking down the why is a bit more effort. If the error output was just "instantiating com_ptr with exceptions disabled, use com_ptr_nothrow instead" it would be super clear and easy to fix the issue without doing any digging. Not a super big issue but seems like it would be a nice to have and potentially save people a few minutes of digging to fix this issue.

**Comment by novakia** [comment]
I don't feel strongly about this, though better errors are always good.

One key to keep in mind if this is pursued is that the solution cannot cause ODR issues when linking a LIB built with exceptions enabled with a LIB built without exceptions enabled into the same binary.  This is commonplace in some binaries.

I suspect that the simple fact that this class is template-based would avoid it in this case, but it could be applicable in others.

**Comment by oldnewthing** [comment]
The problem with the proposal is that it would create ODR violations if one *.cpp file used exceptions and another didn't. This is a common situation for static libraries.

``cpp
// library1.cpp - compiled with exceptions disabled
#include <wil\com.h> // defines err_exception_policy::HResult to raise a static assert
```

``cpp
// library2.cpp - compiled with exceptions enabled
#include <wil\com.h> // defines err_exception_policy::HResult without the static assert
```

If your project uses both libraries, you have an ODR violation if the definition of `err_exception_polcy` or `com_ptr` varied depending on whether exceptions are enabled.

The current mechanism has the important property that you can link together libraries with different exception policies.
